crates/turborepo-ui/src/output.rs

```diff
@@ -0,0 +1,324 @@
+    /// implement Sync as if you want to write to the same sink
+    /// from multiple threads, then you should create a logger for each thread.
+    pub fn logger(&self, behavior: OutputClientBehavior) -> OutputClient<W> {
+        let buffer = match behavior {
+            OutputClientBehavior::Passthrough => None,
+            OutputClientBehavior::InMemoryBuffer | OutputClientBehavior::Grouped => {
+                Some(RefCell::new(Vec::new()))
+            }
+        };
+        let writers = self.writers.clone();
+        OutputClient {
+            behavior,
+            buffer,
+            writers,
+        }
+    }
+}
+
+impl<W: Write> OutputClient<W> {
+    /// A writer that will write to the underlying sink's out writer according
+    /// to this client's behavior.
+    pub fn stdout(&self) -> OutputWriter<W> {
+        OutputWriter {
+            logger: self,
+            destination: Destination::Stdout,
+        }
+    }
+
+    /// A writer that will write to the underlying sink's err writer according
+    /// to this client's behavior.
+    pub fn stderr(&self) -> OutputWriter<W> {
+        OutputWriter {
+            logger: self,
+            destination: Destination::Stderr,
+        }
+    }
+
+    /// Consume the client and flush any bytes to the underlying sink if
+    /// necessary
+    pub fn finish(self) -> io::Result<Option<Vec<u8>>> {
+        let Self {
+            behavior,
+            buffer,
+            writers,
+        } = self;
+        let buffers = buffer.map(|cell| cell.into_inner());
+
+        if matches!(behavior, OutputClientBehavior::Grouped) {
+            let buffers = buffers
+                .as_ref()
+                .expect("grouped logging requires buffer to be present");
+            // We hold the mutex until we write all of the bytes associated for the client
+            // to ensure that the bytes aren't interspersed.
+            let mut writers = writers.lock().expect("lock poisoned");
+            for SinkBytes {
+                buffer,
+                destination,
+            } in buffers
+            {
+                let writer = match destination {
+                    Destination::Stdout => &mut writers.out,
+                    Destination::Stderr => &mut writers.err,
+                };
+                writer.write_all(buffer)?;
+            }
+        }
+
+        Ok(buffers.map(|buffers| {
+            // TODO: it might be worth the list traversal to calculate length so we do a
+            // single allocation
+            let mut bytes = Vec::new();
+            for SinkBytes { buffer, .. } in buffers {
+                bytes.extend_from_slice(&buffer[..]);
+            }
+            bytes
+        }))
+    }
+
+    fn handle_bytes(&self, bytes: SinkBytes) -> io::Result<usize> {
+        if matches!(
+            self.behavior,
+            OutputClientBehavior::InMemoryBuffer | OutputClientBehavior::Grouped
+        ) {
+            let bytes = SinkBytes {
```

Comment:
I'm confused, what is this doing versus just passing `bytes` directly?

Reply:
Is it to convert the buffer to owned?

Reply:
Yes, this is just to change `SinkBytes<'a>` to `SinkBytes<'static>` so we can store it in the in memory buffer.